packages/tools/src/utilities/dynamicVolume/getDataInTime.ts

```diff
@@ -95,19 +94,104 @@ function _getTimePointDataCoordinate(frames, coordinate, volume) {
+  const { imageData: maskImageData } = segmentationVolume;
+  const segScalarData = segmentationVolume.getScalarData();
 
-  for (let i = 0; i < indexArray.length; i++) {
-    const indexValues = [];
-    frames.forEach((frame) => {
-      const activeScalarData = allScalarData[frame];
-      indexValues.push(activeScalarData[indexArray[i]]);
-    });
-    value.push(indexValues);
+  const len = segScalarData.length;
+
+  // Pre-allocate memory for array
+  const nonZeroVoxelIndices = [];
+  nonZeroVoxelIndices.length = len;
+
+  // Get the index of every non-zero voxel in mask
+  let actualLen = 0;
+  for (let i = 0, len = segScalarData.length; i < len; i++) {
+    if (segScalarData[i] !== 0) {
+      nonZeroVoxelIndices[actualLen++] = i;
+    }
   }
-  return value;
+
+  // Trim the array to actual size
+  nonZeroVoxelIndices.length = actualLen;
+
+  const dynamicVolumeScalarDataArray = dynamicVolume.getScalarDataArrays();
+  const values = [];
+  const isSameVolume =
+    dynamicVolumeScalarDataArray[0].length === len &&
+    JSON.stringify(dynamicVolume.spacing) ===
+      JSON.stringify(segmentationVolume.spacing);
+
+  // if the segmentation mask is the same size as the dynamic volume (one frame)
+  // means we can just return the scalar data for the non-zero voxels
+  if (isSameVolume) {
+    for (let i = 0; i < nonZeroVoxelIndices.length; i++) {
+      const indexValues = [];
+      frames.forEach((frame) => {
+        const activeScalarData = dynamicVolumeScalarDataArray[frame];
+        indexValues.push(activeScalarData[nonZeroVoxelIndices[i]]);
+      });
+      values.push(indexValues);
+    }
+
+    return values;
+  }
+
+  // In case that the segmentation mask is not the same size as the dynamic volume (one frame)
+  // then we need to consider each voxel in the segmentation mask and check if it
+  // overlaps with the other volume, and if so we need to average the values of the
+  // overlapping voxels.
+  const callback = ({ pointLPS: segPointLPS, value: segValue }) => {
+    // see if the value is non-zero
+    if (segValue === 0) {
+      // not interested
+      return;
+    }
+
+    // Then for each non-zero voxel in the segmentation mask, we should
+    // again perform the pointInShapeCallback to run the averaging callback
+    // function to get the average value of the overlapping voxels.
+    const overlapIJKMinMax = getVoxelOverlap(
+      dynamicVolume.imageData,
+      dynamicVolume.dimensions,
+      dynamicVolume.spacing,
+      segPointLPS
+    );
+
+    let count = 0;
+    const perFrameSum = new Map();
+
+    // Pre-initialize the Map
+    frames.forEach((frame) => perFrameSum.set(frame, 0));
+
+    const averageCallback = ({ index }) => {
+      for (let i = 0, len = frames.length; i < len; i++) {
+        const value = dynamicVolumeScalarDataArray[i][index];
+        const frame = frames[i];
+        perFrameSum.set(frame, perFrameSum.get(frame) + value);
+      }
+      count++;
```

Comment:
Should leave a comment here. count is representing the number of voxels of the dynamicVolume that represents one voxel in the mask, but the fact that a callback function modifying a global variable is dangerous

Reply:
added